File: R/test_markov_chain.R
#!/usr/bin/env Rscript

library(rbenchmark)


args = commandArgs(TRUE)

# Get the number of iterations n from the command line.
if (length(args) == 0) {
  stop("At least one argument must be supplied (num.iterations).n", call.=FALSE)
} else if (length(args) == 1) {
  n <- as.integer(args[1])
}

f <- function(x) {
  # Math function f(x) to use in the algorithm.
  #
  # Args:
  #   x: Matrix of values to plug in for the function's independent variable.
  # 
  # Returns:
  #   The input matrix after several operations.
  return (exp(sin(x[1]*5) - x[1]*x[1] - x[2]*x[2]))
}

MarkovChain <- function(x, N) {
  # Run the Markov Chain algorithm on the input matrix a given number of times.
  #
  # Args:
  #   x: Matrix of values to plug in for the independent variable.
  #   N: The number of iterations to run the algorithm.
  #
  # Returns:
  #   x: The input matrix after being modified by the algorithm a total of N times.
  p <- f(x)
  for (i in 1:N) {
    x2 <- x + 0.01*rnorm(length(x))
    p2 <- f(x2)
    if (runif(1) < (p2/p)) {
      x <- x2
      p <- p2
    }
  }
  return(x)
}

cat('---------------------------------------\n')
cat('     Markov Chain Calculations: ', n, '\n')
cat('---------------------------------------\n')

benchmark("Markov Calcs" = {
            x <- matrix(0.0, nrow=2, ncol=1)
            MarkovChain(x, n)
          },
          replications = 1,
          columns = c("test", "replications", "elapsed", "relative", "user.self", "sys.self"))

File: Julia/test_markov_chain.jl
using BenchmarkTools
using Printf

# Get the number of iterations from the command line
n, = size(ARGS)
if (n < 1)
    println("Usage: test_markov_chain.jl N")
    println("       ---> Please specify the number of iterations.")
    exit()
end

N = parse(Int, ARGS[1])


"""
    function markov_optimized(x,N)

Operate the Markov chain on two identical inputs x a total of N times.
"""
function markov_optimized(x, N)
    x1, x2 = x
    f(x1, x2) = exp(sin(x1*5) - x1^2 - x2^2)
    p = f(x1, x2)
    for n = 1:N
        y1 = x1 + 0.01*randn(x1)
        y2 = x2 + 0.01*randn(x2)
        q = f(y1, y2)
        if p*rand() < q
            x1 = y1
            x2 = y2
            p = q
        end
    end
end


"""
    function markov(x, y, N)

Operate the Markov chain on two inputs x, y a total of N times.
"""
function markov(x, y, N)
    f(x, y) = exp(sin(x*5) - x^2 - y^2)    
    p = f(x, y)
    for n = 1:N
        x2 = x + 0.01*randn()
        y2 = y + 0.01*randn()
        p2 = f(x2, y2)
        if rand() < p2/p
            x = x2
            y = y2
            p = p2
        end
    end
    x, y
end

# trigger JIT
markov(0.0, 0.0, 10)

println("--------------------------")
println(@sprintf "Markov Chain calculations: %d" N)
println("--------------------------")

@btime markov(0.0, 0.0, N)

#println("--------------------------")
#println(@sprintf "Markov Chain calculations - Opt: %d" N)
#println("--------------------------")
#
@btime markov_optimized([0.0 0.0],N)
